content/blog/2019-11-02-custom-alloc/index.md

```diff
@@ -0,0 +1,206 @@
+ - Combining the first two points, it seems as though dlmalloc would out-perform reap on average if we take out `lcc`.
+ - Almost none of the trends that one can read off of this graph are supported uniformly by all of the points. And Once agin there are only 8 of them.
+
+# From 2002 to 2019
+
+The Lea allocator (Doug Lea's malloc, now referred to as `dlmalloc`) is now the default implementation of linux. The standard general purpose allocator to beat in evaluations is now [jemalloc](http://jemalloc.net/), which seems to be considerably more efficient [^2]. The existence of even better general purpose allocators in some ways strengthens the point made by the paper: there's even less to be gained by writing your own allocator.
+
+On the other hand, custom memory allocation is not quite dead. [Here](https://github.com/mtrebi/memory-allocators)'s a tutorial on how and why custom allocators are helpful, custom allocation still is seen as a potential reason to disregard projects such as [Mesh][], and Emry Berger's [Heap Layers][heapl] project (used to construct reaps in the present paper), which has been much more widely employed, ironically is a framework for building custom allocators, by its own admission:
+
+> "Heap Layers makes it easy to write high-quality custom and general-purpose memory allocators."
+
+
+**What happened to reaps?**
+Though modern memory allocators more effectively put off frees to make them more region-like, "reap" is not a common term for memory allocators --- in fact, only a small handful of the papers that cite this one mention reaps (and all share authors). 
+
+It is interesting that the reap, which one might consider the more substantive, creative contribution to this paper, does not seem to have caught on. In his [2012 blog post](https://emeryblogger.com/2012/10/28/most-influential-oopsla2012/) in which he acknowledges a most influential paper award for this paper, Emery tells people (with extreme brevity and no explanation) to use [his more modern allocator, Hoard](hoard.org), instead of reaps. Even more strangely [the technical paper] that hoard is based on does not include the word "reap", leaving us wondering what happened to them. Still, an implementation lives on in [the legacy portion](https://github.com/emeryberger/Heap-Layers/blob/master/legacy/reap/regionheap.cpp) of the Heap Layers project.
+
+# Conclusions
+
+While this paper seems to have been largely correct in several important ways, many of its elements leave something to be desired, as discussed in the [Empirical Setup section](#experimental-setup). The evaulation seems thorough but unprinicpled and worrylingly oversteps the boundaries of good empiricism with its conclusions; the new algorithm and data structure (reaps) seem very useful and are sold well, but explained phenominally poorly, with only a partial example which doesn't illustrate interesting cases, no theoretical backing, and no explanation of the heuristics that enable its performance. 
```

Comment:
Again, "worrylingly oversteps the boundaries of good empiricism with its conclusions" strikes me as a bit too objectively worded for a subjective judgment. A way to tone it down might be to say "the empirical evaluation has a limited scope and does not use statistics" or something else more objectively true.

Reply:
I agree.